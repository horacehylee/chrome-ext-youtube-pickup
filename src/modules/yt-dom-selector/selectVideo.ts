import * as select from "select-dom";

export const selectVideo = (videoId: string): HTMLAnchorElement => {
  let _videoId = videoId;
  if (/^[0-9]/.test(_videoId)) {
    _videoId = _videoId.substring(1);
  }
  return select(`a[href*=${_videoId}][id=thumbnail]`);
};
