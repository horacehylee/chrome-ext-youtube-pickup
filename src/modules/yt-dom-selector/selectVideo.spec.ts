import { readFile } from "fs";
import { promisify } from "bluebird";
import { selectVideo } from "./selectVideo";
const readFileAsync = promisify(readFile);

interface SelectVideoTestOption {
  description: string;
  sampleFileName: string;
  videoId: string;
}

const buildTests = (options: SelectVideoTestOption[]) => {
  return options.map(({ description, sampleFileName, videoId }) => {
    it(description, async () => {
      const buffer = await readFileAsync(
        `${__dirname}/../../../sample/${sampleFileName}`
      );
      document.body.innerHTML = buffer.toString();
      const element = selectVideo(videoId);

      expect(element).toBeDefined();
      expect(element).toBeInstanceOf(HTMLAnchorElement);
      expect(element).toMatchSnapshot();
    });
  });
};

describe("selectVideo", () => {
  buildTests([
    {
      description: "should select videoId element",
      sampleFileName: "ytVideoElement.html",
      videoId: "sLgKXX0FrJo"
    },
    {
      description: "should select video element with time value in url",
      sampleFileName: "ytVideoElement_withTime.html",
      videoId: "xu6KLURIuh8"
    },
    {
      description:
        "should select video element with number as the first character",
      sampleFileName: "ytVideoElement_numberAsFirstChar.html",
      videoId: "2LifjpxpIwU"
    }
  ]);
});
