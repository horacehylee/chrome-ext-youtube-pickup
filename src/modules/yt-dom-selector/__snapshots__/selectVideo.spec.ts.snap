// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectVideo should select video element with number as the first character 1`] = `
<a
  aria-hidden="true"
  class="yt-simple-endpoint inline-block style-scope ytd-thumbnail"
  href="/watch?v=2LifjpxpIwU"
  id="thumbnail"
  rel="null"
  tabindex="-1"
>
  
                      
  <yt-img-shadow
    class="style-scope ytd-thumbnail no-transition"
    loaded=""
    style="background-color: transparent;"
  >
    
                        
    <img
      alt=""
      class="style-scope yt-img-shadow"
      id="img"
      src="https://i.ytimg.com/vi/2LifjpxpIwU/hqdefault.jpg?sqp=-oaymwEZCNACELwBSFXyq4qpAwsIARUAAIhCGAFwAQ==&rs=AOn4CLB0HNYDXawpIC-8pZVuwgWck_K4FQ"
      width="246"
    />
    
                      
  </yt-img-shadow>
  

                      
  <div
    class="style-scope ytd-thumbnail"
    id="overlays"
  >
    
                        
    <ytd-thumbnail-overlay-time-status-renderer
      class="style-scope ytd-thumbnail"
      overlay-style="DEFAULT"
    >
      
                          
      <span
        aria-label="2 hours, 5 minutes"
        class="style-scope ytd-thumbnail-overlay-time-status-renderer"
      >
        
                            2:05:41
                          
      </span>
      
                        
    </ytd-thumbnail-overlay-time-status-renderer>
    
                      
  </div>
  
                      
  <div
    class="style-scope ytd-thumbnail"
    id="mouseover-overlay"
  />
  
                      
  <div
    class="style-scope ytd-thumbnail"
    id="hover-overlays"
  />
  
                    
</a>
`;

exports[`selectVideo should select video element with time value in url 1`] = `
<a
  aria-hidden="true"
  class="yt-simple-endpoint inline-block style-scope ytd-thumbnail"
  href="/watch?v=xu6KLURIuh8&t=29s"
  id="thumbnail"
  rel="null"
  tabindex="-1"
>
  
                      
  <yt-img-shadow
    class="style-scope ytd-thumbnail no-transition"
    loaded=""
    style="background-color: transparent;"
  >
    
                        
    <img
      alt=""
      class="style-scope yt-img-shadow"
      id="img"
      src="https://i.ytimg.com/vi/xu6KLURIuh8/hqdefault.jpg?sqp=-oaymwEZCNACELwBSFXyq4qpAwsIARUAAIhCGAFwAQ==&rs=AOn4CLCBjK3OTHsBnJqdrZVSB2BWJamVhg"
      width="246"
    />
    
                      
  </yt-img-shadow>
  

                      
  <div
    class="style-scope ytd-thumbnail"
    id="overlays"
  >
    
                        
    <ytd-thumbnail-overlay-resume-playback-renderer
      class="style-scope ytd-thumbnail"
    >
      
                          
      <div
        class="style-scope ytd-thumbnail-overlay-resume-playback-renderer"
        id="progress"
        style="width: 13%;"
      />
      
                        
    </ytd-thumbnail-overlay-resume-playback-renderer>
    
                        
    <ytd-thumbnail-overlay-time-status-renderer
      class="style-scope ytd-thumbnail"
      overlay-style="DEFAULT"
    >
      
                          
      <span
        aria-label="3 minutes, 42 seconds"
        class="style-scope ytd-thumbnail-overlay-time-status-renderer"
      >
        
                            3:42
                          
      </span>
      
                        
    </ytd-thumbnail-overlay-time-status-renderer>
    
                      
  </div>
  
                      
  <div
    class="style-scope ytd-thumbnail"
    id="mouseover-overlay"
  />
  
                      
  <div
    class="style-scope ytd-thumbnail"
    id="hover-overlays"
  />
  
                    
</a>
`;

exports[`selectVideo should select video elements with more than one numbers as first characters 1`] = `
<a
  aria-hidden="true"
  class="yt-simple-endpoint inline-block style-scope ytd-thumbnail"
  href="/watch?v=5700buJP-Qw"
  id="thumbnail"
  rel="null"
  tabindex="-1"
>
  
                    
  <yt-img-shadow
    class="style-scope ytd-thumbnail no-transition"
    loaded=""
    style="background-color: transparent;"
  >
    
                      
    <img
      alt=""
      class="style-scope yt-img-shadow"
      id="img"
      src="https://i.ytimg.com/vi/5700buJP-Qw/hqdefault.jpg?sqp=-oaymwEZCNACELwBSFXyq4qpAwsIARUAAIhCGAFwAQ==&rs=AOn4CLDNKKwRb5q-YboSSsWp89ClSJJ0Ug"
      width="246"
    />
    
                    
  </yt-img-shadow>
  

                    
  <div
    class="style-scope ytd-thumbnail"
    id="overlays"
  >
    
                      
    <ytd-thumbnail-overlay-time-status-renderer
      class="style-scope ytd-thumbnail"
      overlay-style="DEFAULT"
    >
      
                        
      <span
        aria-label="9 minutes, 29 seconds"
        class="style-scope ytd-thumbnail-overlay-time-status-renderer"
      >
        
                          9:29
                        
      </span>
      
                      
    </ytd-thumbnail-overlay-time-status-renderer>
    
                    
  </div>
  
                    
  <div
    class="style-scope ytd-thumbnail"
    id="mouseover-overlay"
  />
  
                    
  <div
    class="style-scope ytd-thumbnail"
    id="hover-overlays"
  />
  
                  
</a>
`;

exports[`selectVideo should select videoId element 1`] = `
<a
  aria-hidden="true"
  class="yt-simple-endpoint inline-block style-scope ytd-thumbnail"
  href="/watch?v=sLgKXX0FrJo"
  id="thumbnail"
  rel="null"
  tabindex="-1"
>
  
                            
  <yt-img-shadow
    class="style-scope ytd-thumbnail no-transition"
    loaded=""
    style="background-color: transparent;"
  >
    
                              
    <img
      alt=""
      class="style-scope yt-img-shadow"
      id="img"
      src="https://i.ytimg.com/vi/sLgKXX0FrJo/hqdefault.jpg?sqp=-oaymwEZCNACELwBSFXyq4qpAwsIARUAAIhCGAFwAQ==&rs=AOn4CLAhX9R61CoQRp-RQI2eNVDae-nO8Q"
      width="246"
    />
    
                            
  </yt-img-shadow>
  

                            
  <div
    class="style-scope ytd-thumbnail"
    id="overlays"
  >
    
                              
    <ytd-thumbnail-overlay-time-status-renderer
      class="style-scope ytd-thumbnail"
      overlay-style="DEFAULT"
    >
      
                                
      <span
        aria-label="9 minutes, 48 seconds"
        class="style-scope ytd-thumbnail-overlay-time-status-renderer"
      >
        
                                  9:48
                                
      </span>
      
                              
    </ytd-thumbnail-overlay-time-status-renderer>
    
                            
  </div>
  
                            
  <div
    class="style-scope ytd-thumbnail"
    id="mouseover-overlay"
  />
  
                            
  <div
    class="style-scope ytd-thumbnail"
    id="hover-overlays"
  />
  
                          
</a>
`;
